SMPyBandits/Policies/RAWUCB.py

```diff
@@ -0,0 +1,109 @@
+from __future__ import division, print_function  # Python 2 compatibility
+
+__author__ = "Julien Seznec"
+__version__ = "0.1"
+
+from math import sqrt, log
+import numpy as np
+import pandas as pd
+
+np.seterr(divide='ignore')  # XXX dangerous in general, controlled here!
+
+try:
+    from .BasePolicy import BasePolicy
+    from .IndexPolicy import IndexPolicy
+    from .FEWA import FEWA, EFF_FEWA
+    from .kullback import klucbBern
+except ImportError:
+    from BasePolicy import BasePolicy
+    from FEWA import FEWA, EFF_FEWA
+    from kullback import klucbBern
+    from IndexPolicy import IndexPolicy
+
+
+class EFF_RAWUCB(EFF_FEWA):
+    """
+    Efficient mechanism as described in [Seznec et al.,  2019a] (m=2) and [Seznec et al.,  2019b] (m<=2)
+    Algorithm Rotting Adaptive Window Upper Confidence Bound (RAW-UCB) [Seznec et al.,  2019b]
+    We use the confidence level \delta_t = \frac{1}{t^\alpha}.
+    """
+
+    def choice(self):
+        not_selected = np.where(self.pulls == 0)[0]
+        if len(not_selected):
+            return not_selected[0]
+        self.ucb = self._compute_ucb()
+        return np.nanmin(self.ucb, axis=1).argmax()
```

Comment:
That's a smart way to compute the chosen arm, I like the `np.nanmin().argmax()`. I would suggest to use function instead of methods, so `np.argmax(np.nanmin(self.ucb, axis=1))`.

Reply:
what is the benefit?

Reply:
Imagine I now decide to use another algorithm instead of UCB, one which maybe compute its UCB only once in a while. The `np.nanmin(self.ucb, axis=1)` could be computed elsewhere, and maybe stored as a Python list rather than a numpy array ? (if its your intern who is coding…)
Using `array_foo.argmax()` only works when `array_foo` is a numpy array, but not if it becomes a list (or any other iterable), on which `np.argmax(…)` works.

Another reason is that usually the text editors recognize wrapped functions more cleanly than methods called on an object *which is the result of a computation*!